### PR TITLE
Mutable module improvement

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 1]
```diff
@@ -189,6 +194,9 @@ def __init__(
             "hardware_compatible": hardware_compatible,
             "timing_cache_path": timing_cache_path,
         }
+        self.arg_dynamic_shapes: Optional[tuple[Any]] = None
+        self.kwarg_dynamic_shapes: Optional[dict[Any, Any]] = None
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:198]: Do we need all three of these members? Isnt total dynamic shape easily construct-able from `arg_dynamic_shapes` and `kwarg_dynamic_shape`?

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 2]
```diff
@@ -203,6 +211,31 @@ def __init__(
         )
         self.init_finished = True
 
+    def set_dynamic_shape_hint(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:214]: Please add docstrings to all user facing functions

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 3]
```diff
@@ -203,6 +211,31 @@ def __init__(
...
+        self,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:215]: Will this be applicable for static shape as well? Would this be a usecase that would be useful?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:215]: Otherwise we should call it `set_expected_shape_range` or something

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py (hunk 1 above)
Comment (narendasan) [reply py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:198]: In general we want to minimize state to prevent desync issues

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 4]
```diff
@@ -306,14 +340,26 @@ def compile(self) -> None:
...
+                    self.kwarg_inputs, kwargs, dynamic_shapes=self.kwarg_dynamic_shapes
+                )
+            ):
+                logger.info("Input change detected.")
+                self.refit_state.set_state(RefitFlag.NEEDS_RECOMPILE)
+                self.store_inputs(args, kwargs)
+        except DynamicShapeOutOfRangeException as e:
             logger.info("Input change detected.")
+            logger.warning(e)
+            logger.warning("Recompiling the engine with static shape")
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:359]: Make this warning more verbose. Something like `Provided inputs are outside the set expected shape range, recompiling module for the provided input shapes (static)`

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 5]
```diff
@@ -306,14 +340,26 @@ def compile(self) -> None:
...
-        if (
-            not self.arg_inputs
-            or not MutableTorchTensorRTModule.check_inputs_equal(self.arg_inputs, args)
-            or not MutableTorchTensorRTModule.check_inputs_equal(
-                self.kwarg_inputs, kwargs
-            )
-        ):
+        try:
+            if (
+                not self.arg_inputs
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:345]: This condition is a bit confusing to read

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py [hunk 6]
```diff
@@ -436,35 +482,68 @@ def __setattr__(self, name: str, value: Any) -> None:
     def check_inputs_equal(
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:482]: Should this be renamed `check_inputs_in_range`?

File: tests/py/dynamo/runtime/test_mutable_torchtrt_module.py [hunk 7]
```diff
@@ -36,6 +36,126 @@ def test_check_output_equal():
...
+        "c": {"a": torch.rand(10, 30), "b": [torch.rand(10, 30), torch.rand(5, 20)]},
+    }
+
+    kwargs = torch_trt.MutableTorchTensorRTModule.process_kwarg_inputs(kwargs_5)
+    trt_gm._validate_inputs(*inputs_5, **kwargs_5)
+    assertions.assertTrue(
+        trt_gm.refit_state.get_state() == RefitFlag.NEEDS_RECOMPILE,
+        msg=f"Dynamic shape support is not correct.",
+    )
+    trt_gm(*inputs_5, **kwargs_5)
```
Comment (narendasan) [inline tests/py/dynamo/runtime/test_mutable_torchtrt_module.py:156]: Why do we run these after the assert?

File: examples/dynamo/mutable_torchtrt_module_example.py [hunk 8]
```diff
@@ -103,7 +101,7 @@
     # Standard Huggingface LoRA loading procedure
     pipe.load_lora_weights(
         "stablediffusionapi/load_lora_embeddings",
-        weight_name="moxin.safetensors",
+        weight_name="all-disney-princess-xl-lo.safetensors",
         adapter_name="lora1",
     )
     pipe.set_adapters(["lora1"], adapter_weights=[1])
```
Comment (narendasan) [inline examples/dynamo/mutable_torchtrt_module_example.py:107]: Can you add a example for the shape hint (could be separate and simpler)

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py (hunk 1 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:198]:
```
self.total_dynamic_shape = self.kwarg_dynamic_shapes.copy()
signature = list(
    inspect.signature(self.original_model.forward).parameters.keys()
)
for i, arg in enumerate(self.arg_dynamic_shapes):
    self.total_dynamic_shape[signature[i]] = arg
```
This is how to construct a total shape from those two. Since the arg shape is a tuple and does not have the argument name as key value, I have to iterate the signature to fetch them. Do you think we should run this everytime we recompile?

File: py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py (hunk 6 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/runtime/_MutableTorchTensorRTModule.py:482]: Can I call it check_inputs_shape? This function handles both static and dynamic scenario and there is no "range" in static shape.

File: tests/py/dynamo/runtime/test_mutable_torchtrt_module.py [hunk 9]
```diff
@@ -36,6 +36,126 @@ def test_check_output_equal():
...
+    kwargs_2 = {
+        "b": torch.rand(9, 30),
+        "c": {"a": torch.rand(10, 30), "b": [torch.rand(10, 30), torch.rand(5, 20)]},
+    }
+
+    kwargs = torch_trt.MutableTorchTensorRTModule.process_kwarg_inputs(kwargs_2)
+    trt_gm._validate_inputs(*inputs_2, **kwargs_2)
+    assertions.assertTrue(
+        trt_gm.refit_state.get_state() == RefitFlag.LIVE,
+        msg=f"Dynamic shape support is not correct.",
```
Comment (peri044) [inline tests/py/dynamo/runtime/test_mutable_torchtrt_module.py:120]: Modify this message to include information about which inputs failed (like `inputs_2` etc) to distinguish from other assert messages below

File: tests/py/dynamo/runtime/test_mutable_torchtrt_module.py [hunk 10]
```diff
@@ -36,6 +36,126 @@ def test_check_output_equal():
...
 
+@pytest.mark.unit
+def test_check_input_shape_dynamic():
+    torch.manual_seed(0)
+    a = {
+        "a": torch.rand(10, 3),
+        "b": [torch.rand(10, 30), torch.rand(5, 5)],
+        "c": {"a": torch.rand(10, 30), "b": [torch.rand(10, 30), torch.rand(5, 5)]},
+    }
+    torch.manual_seed(0)
```
[the remark this answers is not in the document]
Comment (cehongwang) [reply tests/py/dynamo/runtime/test_mutable_torchtrt_module.py:58]: This is not redundant. If we don't reset the seed the output is different

File: examples/dynamo/mutable_torchtrt_module_example.py [hunk 11]
```diff
@@ -113,3 +110,45 @@
...
+    "c": {"a": torch.rand(10, 30).to(device), "b": torch.rand(10, 30).to(device)},
+}
+dim_0 = torch.export.Dim("dim", min=1, max=50)
+dim_1 = torch.export.Dim("dim", min=1, max=50)
+dim_2 = torch.export.Dim("dim2", min=1, max=50)
+args_dynamic_shapes = ({1: dim_1}, {0: dim_0})
+kwarg_dynamic_shapes = {
+    "c": {"a": {}, "b": {0: dim_2}},
+}
+# Export the model first with custom dynamic shape constraints
```
Comment (narendasan) [inline examples/dynamo/mutable_torchtrt_module_example.py:143]: Add some more detail about what the semantics of the dynamic shape hint is

**Review by narendasan — COMMENTED** [review]
Do you have the persistent cache example?

**Comment by cehongwang** [comment]
> Do you have the persistent cache example?

I talked to Boris and it seems like save and load is what he is looking for. I am adding engine caching example to MutableTorchTensorRTModule as well.

File: tests/py/dynamo/runtime/test_mutable_torchtrt_module.py [hunk 12]
```diff
@@ -35,6 +35,160 @@ def test_check_output_equal():
...
+        "a": torch.rand(10, 30),
+        "b": [torch.rand(10, 30), torch.rand(5, 5)],
+        "c": {"a": torch.rand(10, 30), "b": [torch.rand(10, 30), torch.rand(5, 5)]},
+    }
+
+    dim = torch.export.Dim("dim", min=1, max=50)
+    dynamic_shape = {"a": {1: dim}, "b": [{}, {}], "c": {"a": {}, "b": [{}, {}]}}
+    assertions.assertFalse(
+        torch_trt.MutableTorchTensorRTModule._check_inputs_shape(a, b),
+        msg=f"test_check_output_equal is not correct.",
```
Comment (peri044) [inline tests/py/dynamo/runtime/test_mutable_torchtrt_module.py:69]: test_check_input_shape_dynamic

File: tests/py/dynamo/runtime/test_mutable_torchtrt_module.py [hunk 13]
```diff
@@ -35,6 +35,160 @@ def test_check_output_equal():
...
+    )
+    assertions.assertTrue(
+        torch_trt.MutableTorchTensorRTModule._check_inputs_shape(a, b, dynamic_shape),
+        msg=f"test_check_output_equal is not correct.",
```
Comment (peri044) [inline tests/py/dynamo/runtime/test_mutable_torchtrt_module.py:73]: test_check_input_shape_dynamic

**Review by peri044 — COMMENTED** [review]
Mostly LGTM. Added minor comments
